NuRadioMC/SignalProp/analyticraytracing.py

```diff
@@ -1833,11 +1855,385 @@ def get_path(self, iS, n_points=1000):
+    
+    def on_sky_birefringence(self, theta, phi, polarization):
+
+        """
+        Function for the normalized e-field vector from cartesian to spherical coordinates.
+        The function does the same as the following radiotool functions, only faster:             
+        from radiotools import coordinatesystems
+        cs = coordinatesystems.cstrafo(theta, phi)
+        sky = cs.transform_from_ground_to_onsky(p)
+
+        Parameters
+        ----------
+        theta: float
+            zenith angle of the propagation direction
+        phi: float
+            azimuth angle of the propagation direction
+        polarization: np.array([px, py, pz])
+            normalized e-field vector in cartesian coordinates
+
+        Returns
+        -------
+        output format: numpy.array
+            np.array([p_radial_1, p_theta_1, p_phi_1])
+        meaning       
+            normalized e-field vector in spherical coordinates
+        """
+
+        transform = np.array([  [np.sin(theta) * np.cos(phi) , np.sin(theta) * np.sin(phi)   , np.cos(theta)    ], 
+                                [np.cos(theta) * np.cos(phi) , np.cos(theta) * np.sin(phi)   , - np.sin(theta)  ], 
+                                [- np.sin(phi)               , np.cos(phi)                   , 0                ]       ])
+
+        return transform.dot(polarization)
+
+    def get_pulse_propagation_birefringence(self, pulse, samp_rate, i_solution, bire_model = 'southpole_A'):
+        
+        """
+        Function for the time trace propagation according to the polarization change due to birefringence. 
+        The trace propagation is explained in this paper: https://link.springer.com/article/10.1140/epjc/s10052-023-11238-y
+
+        Parameters
+        ----------
+        pulse: np.ndarray
+            3d array with the frequency spectrum of np.array([eR, eTheta, ePhi]), usually provided by the apply_propagation_effects function
+        samp_rate: float
+            sampling rate of the time traces
+        i_solution: int
+            choose which ray-tracing solution should be propagated
+        bire_model: string
+            choose the interpolation to fit the measured refractive index data
+            options include (A, B, C, D, E) description can be found under: NuRadioMC/NuRadioMC/utilities/birefringence_models/model_description
+
+        Returns
+        -------
+
+        final pulse: numpy.array([eR, eTheta, ePhi])
+            [0] - eR        - final frequency spectrum of the radial component - not altered by the function
+            [1] - eTheta    - final frequency spectrum of the theta component
+            [2] - ePhi      - final frequency spectrum of the phi component
+        """
+
+        t_fast = base_trace.BaseTrace()
+
+        ice_n = self.__medium
+        ice_birefringence = medium.get_ice_model('birefringence_medium')
+        ice_birefringence.__init__(bire_model)
+
+        acc = int(self.get_path_length(i_solution))
```

Comment:
implicit assumption that the base unit is meters. Fix this by adding a `/units.m`

Reply:
Done